### PR TITLE
Place vtable slots deemed final by analysis into sealed vtable

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Virtual method table slots typically go into vtable - a variable-length part of the `MethodTable` data structure. Virtual method table slots are always inherited by `MethodTable` of the derived class. This means that vtable slots that are introduced in a class that has many derived classes can be quite expensive. One example are arrays - array types inherit all virtual methods from `System.Array` and there's many arrays.

We therefore have an optimization - if a virtual method is newly introduced, but it's sealed at the same time, we place it into a separate data structure - a "sealed vtable". Sealed vtable slots do not get inherited by derived classes. Because they don't get inherited, we also need to make sure all non-interface calls into this virtual method get devirtualized (virtual dispatch needs to be aware of the special lookup that happens for sealed slots and only interface resolution is aware of how to do that).

This PR makes this optimization kick in more often - not just when method is `virtual sealed` in metadata (which effectively only happens for non-virtual method in C# implementing an interface), but also when it's virtual and nothing else overrides it in the program.

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/MethodExtensions.cs [hunk 1]
```diff
@@ -82,25 +87,43 @@ public static string GetUnmanagedCallersOnlyExportName(this EcmaMethod This)
...
+            // Implementations of static virtual methods go into the sealed vtable.
+            if (method.Signature.IsStatic)
                 return true;
 
-            // Implementations of static virtual method also go into the sealed vtable.
-            // Again, we don't let that happen for interface methods because the slot numbers are only logical,
-            // not physical.
-            if (method.Signature.IsStatic && !isInterfaceMethod)
+            // If the owning type is already considered sealed, there's little benefit in placing the slots
+            // in the sealed vtable: the sealed vtable has these properties:
```
Comment (jkotas) [inline src/coreclr/tools/Common/Compiler/MethodExtensions.cs:105]: What are the downsides of keeping this simple and placing them in the sealed vtable? Are you trying to save the extra pointer to the sealed vtable?

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/Compiler/MethodExtensions.cs:105]: Yep, the extra pointer to the vtable. The old logic was just `method.IsFinal && method.IsNewSlot` so we didn't even look whether the type is sealed. Now we end up doing this for a lot more methods, often on valuetypes where it makes little sense.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**